src/algorithms/calorimetry/TrackClusterMergeSplitter.cc

```diff
@@ -0,0 +1,462 @@
+
+        // increment sum and print debugging message
+        eTotal += eContrib;
+        trace("Added hit: eHit = {}, weight = {}, eContrib = {}, running total = {}", hit.getEnergy(), weight, eContrib, eTotal);
+
+        // set hit one-to-many/vector relations
+        merged_clust.addToHits( hit );
+        merged_clust.addToWeights( weight );
+
+      }  // end hit loop
+      debug("Merged input cluster {} into output cluster {}", old_clust.getObjectID().index, merged_clust.getObjectID().index);
+
+    }  // end of cluster loop
+
+  }  // end 'merge_clusters(std::vector<edm4eic::Cluster>&, edm4eic::MutableCluster&)'
+
+
+
+
+
+  // --------------------------------------------------------------------------
+  //! Copy cluster onto new one
+  // --------------------------------------------------------------------------
+  void TrackClusterMergeSplitter::copy_cluster(
+    const edm4eic::ProtoCluster& old_clust,
+    edm4eic::MutableProtoCluster& new_clust
+  ) const {
+
+    // set one-to-many relations
+    for (auto hit : old_clust.getHits()) {
+      new_clust.addToHits( hit );
+    }
+
+    // set vector members
+    for (auto weight : old_clust.getWeights()) {
+      new_clust.addToWeights( weight );
+    }
+    debug("Copied input cluster {} onto output cluster {}", old_clust.getObjectID().index, new_clust.getObjectID().index);
+
+  }  // end 'copy_cluster(edm4eic::Cluster&, edm4eic::MutableCluster&)'
+
+
+
+  // --------------------------------------------------------------------------
+  //! Grab current energy of protocluster
+  // --------------------------------------------------------------------------
+  float TrackClusterMergeSplitter::get_cluster_energy(const edm4eic::ProtoCluster& clust) const {
+
+    float eClust = 0.;
+    for (auto hit : clust.getHits()) {
+      eClust += hit.getEnergy();
+    }
+    return eClust / m_cfg.sampFrac;
+
+  }  // end 'get_cluster_energy(edm4eic::ProtoCluster&)'
+
+
+
+  // --------------------------------------------------------------------------
+  //! Get current center of protocluster
+  // --------------------------------------------------------------------------
+  edm4hep::Vector3f TrackClusterMergeSplitter::get_cluster_position(const edm4eic::ProtoCluster& clust) const {
+
+    // grab total energy
+    const float eClust = get_cluster_energy(clust) * m_cfg.sampFrac;
+
+    // calculate energy-weighted center
+    float wTotal = 0.;
+    edm4hep::Vector3f position(0., 0., 0.);
+    for (auto hit : clust.getHits()) {
+
+      // calculate weight
+      float weight = hit.getEnergy() / eClust;
+      wTotal += weight;
+
+      // update cluster position
+      position = position + (hit.getPosition() * weight);
+    }
+
+    float norm = 1.;
+    if (wTotal == 0.) {
+      warning("Total weight of 0 in position calculation!");
+    } else {
+      norm = wTotal;
+    }
+    return position / norm;
+
+  }  // end 'get_cluster_position(edm4eic::ProtoCluster&)'
```

Comment:
This is something I've been thinking about for a bit now... I was initially planning on using `edm4eic::Cluster` as the input, but I opted for `edm4eic::Protocluster` instead since I felt that I would be duplicating more of ClusterRecoCoG's functionality if I stuck with just clusters (namely rerunning the cluster shape calculations, the CoG calculation, and handling the associations).

The trade-off, though, is that I lose the position as a member of the object. Which is why it gets duplicated here...

Reply:
I've been thinking that routines like this, the cluster shape calculations, and similar that could get used across multiple algorithms might be useful to store somewhere centralized like an `edm4eic::utils`. @wdconinc and I briefly chatted about at the UGM...

But, of course, this is more just food for thought since this would be outside the scope of this PR 🤷